### PR TITLE
add span based metrics

### PR DESCRIPTION
**Pull request by DeNeutoy**
This adds span based metrics to the SRL model. I implemented it as a `Metric` but it doesn't fit the API perfectly, because it requires a `tag_vocab` to be passed to it, making it hard to build from params. Returning the metrics not as a `dict` is also hard because there are possibly a variable number of them and there will be quite a few. Maybe we should switch the other metrics to be `dicts` too - I think this is the easiest solution. 

Although it's not perfect, it does mean we can get a much better measurement of the SRL model performance - once this is in we are ready to start training a few big models, I think. 

Problems/considerations Summary:
- This doesn't count continuations and references as part of the previous span they are referencing/continuing, which I think is fairly minor.
- Having to pass a vocab to the metric makes it hard to build from params.
- Passing the computed metrics back to the model without using a dict is messy, currently breaking type checking.

### CODE REVIEW

File: allennlp/training/metrics.py [hunk 1]
```diff
@@ -288,3 +290,196 @@ def reset(self):
...
+                 label_vocabulary: Dict[int, str],
+                 ignore_classes: List[int] = None) -> None:
+        """
+        Parameters
+        ----------
+        label_vocabulary : Dict[int, str], required.
+            A mapping from integer class labels to string labels. This metric
+            assumes that a BIO format is used in which the labels are of the format:
+            ["B-LABEL", "I-LABEL"].
+        ignore_classes : List[int], optional.
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Because you have the vocabulary, would it make sense to pass these in as strings?  Oh, I guess you're just taking the `Dict[int, str]`, so it'd be obnoxious to find the strings...  Maybe this is ok.  Makes me wonder if we should have designed the `Vocab` differently, so each namespace is a class, and we could have just taken that here, and what is now the `Vocabulary` object is just a dictionary mapping strings to these namespace vocab objects...

File: allennlp/training/metrics.py [hunk 2]
```diff
@@ -288,3 +290,196 @@ def reset(self):
...
+        mask: ``torch.Tensor``, optional (default = None).
+            A masking tensor the same size as ``gold_labels``.
+        """
+        if mask is None:
+            mask = torch.ones(gold_labels.size())
+        # If you actually passed in Variables here instead of Tensors, this will be a huge memory
+        # leak, because it will prevent garbage collection for the computation graph.  We'll ensure
+        # that we're using tensors here first.
+        if isinstance(predictions, Variable):
+            predictions = predictions.data.cpu()
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Why `.cpu()`?  Because of the loop stuff you have?  Does it make a difference?

File: allennlp/training/metrics.py [hunk 3]
```diff
@@ -288,3 +290,196 @@ def reset(self):
...
+        if isinstance(mask, Variable):
+            mask = mask.data.cpu()
+
+        num_classes = predictions.size(-1)
+        if (gold_labels >= num_classes).any():
+            raise ConfigurationError("A gold label passed to SpanBasedF1Measure contains an "
+                                     "id >= {}, the number of classes.".format(num_classes))
+
+        sequence_lengths = get_lengths_from_binary_sequence_mask(mask)
+        argmax_predictions = torch.topk(predictions, 1, -1)[1].float().squeeze(-1)
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Can't you just use `torch.max()` here?

File: allennlp/training/metrics.py [hunk 4]
```diff
@@ -288,3 +290,196 @@ def reset(self):
...
+        Parameters
+        ----------
+        tag_sequence : List[int], required.
+            The integer class labels for a sequence.
+
+        Returns
+        -------
+        spans : Set[Tuple[Tuple[int, int], str]]
+            The typed, extracted spans from the sequence, in the format ((span_start, span_end), label).
+            Note that the label _does not_ contain any BIO tag prefixes.
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Replace `_` with ``` ` ```

File: allennlp/training/metrics.py [hunk 5]
```diff
@@ -288,3 +290,196 @@ def reset(self):
         self._true_negatives = 0.0
         self._false_positives = 0.0
         self._false_negatives = 0.0
+
+
+@Metric.register("conll_span_f1")
+class ConllSpanBasedF1Measure(Metric):
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: I'd remove `Conll` from this name - Cristian could use this for his F1 metric, for instance.  It's just a general span-based F1 for BIO-tagged data.

File: allennlp/training/metrics.py [hunk 6]
```diff
@@ -288,3 +290,196 @@ def reset(self):
...
+        spans = set()
+        span_start = 0
+        span_end = 0
+        active_conll_tag = None
+        for index, integer_tag in enumerate(tag_sequence):
+            # Actual BIO tag.
+            string_tag = self._label_vocabulary[integer_tag]
+            bio_tag = string_tag[0]
+            conll_tag = string_tag[2:]
+            if integer_tag in self._ignore_classes:
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: Having now seen this logic, I think it'd be better to just take a set of `conll_tag` _strings_ as input to the constructor.  If the caller forgot to provide the index for "O" here, for example, this logic would be incorrect, but that's a really easy case to handle here.  If you do it this way, the caller can just say "ignore 'V' and 'ARG-LOC' spans", and it's easy, instead of having to query the dictionary for `B-V`, `I-V`, `B-ARG-LOC`, `I-ARG-LOC`, and `O`.  Looks like in your instantiation of the metric in the SRL model, you also forgot `I-V`.  Looking at the logic, I'm having a hard time telling if that will cause a bug or not.  Easier to reason about if the things you're ignoring are `conll_tags` instead of integer labels.

I think it would also make this piece of code easier to extend to the `C-` and other cases later, if we wanted to.

File: allennlp/models/semantic_role_labeler.py [hunk 7]
```diff
@@ -175,6 +183,14 @@ def tag(self, text_field: TextField, verb_indicator: IndexField) -> Dict[str, An
 
         return {"tags": tags, "class_probabilities": predictions.numpy()}
 
+    def get_metrics(self, reset: bool = False):
+        # TODO(Mark): This won't play well with the metrics in Trainer, see how Matt wants it.
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: Ah, having now seen the metric code, no it's not.  I'd flatten out the dictionary either in the `Metric` class or here (probably in the `Metric` class), so there's only one `Dict[str, float]`.

File: allennlp/training/metrics.py [hunk 8]
```diff
@@ -288,3 +290,196 @@ def reset(self):
         self._true_negatives = 0.0
         self._false_positives = 0.0
         self._false_negatives = 0.0
+
+
+@Metric.register("conll_span_f1")
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: On second thought: another option is to just allow all metrics to take the `Vocabulary` in their `from_params` method.  If you do that, along with switching the ignored classes from ints to strings, it'd be relatively straightforward to construct this `from_params`.  That's probably the better way to go.

File: allennlp/training/metrics.py [hunk 9]
```diff
@@ -288,3 +290,196 @@ def reset(self):
...
+        A Dict per label containing following the span based metrics:
+        precision : float
+        recall : float
+        f1-measure : float
+
+        Additionally, an ``overall`` key is included, which provides the precision,
+        recall and f1-measure for all spans.
+        """
+        all_tags = set()  # type: Set[str]
+        all_tags.update(self._true_positives)
```
Comment (matt-gardner) [inline allennlp/training/metrics.py]: This is confusing to read - maybe use `self._true_positives.keys()`?

File: tests/training/metrics_test.py [hunk 10]
```diff
@@ -211,3 +213,89 @@ def test_f1_measure_works_for_sequences(self):
...
+        tag_vocab = self.vocab.get_index_to_token_vocabulary("tags")
+        metric = ConllSpanBasedF1Measure(tag_vocab, ignore_classes=[o_index])
+
+        tag_sequence = ["O", "B-ARG1", "I-ARG1", "O", "B-ARG2", "I-ARG2", "B-ARG1", "B-ARG2"]
+        indices = [self.vocab.get_token_index(x, "tags") for x in tag_sequence]
+        spans = metric._extract_spans(indices)
+        assert spans == {((1, 2), "ARG1"), ((4, 5), "ARG2"), ((6, 6), "ARG1"), ((7, 7), "ARG2")}
+
+        # Check that invalid BIO sequences are handled as non-spans.
+        tag_sequence = ["O", "B-ARG1", "I-ARG1", "O", "I-ARG1", "B-ARG2", "I-ARG2", "B-ARG1", "I-ARG2"]
```
Comment (matt-gardner) [inline tests/training/metrics_test.py]: Hmm, seems like the model should be penalized in precision if it gives invalid BIO sequences.  I'd probably go the other way and count invalid tags like this as predictions.  Otherwise, if I have a predicted tag sequence that differs from a gold tag sequence only by replacing an "O" somewhere with an "I-ARG1", I'll get a perfect F1 score, when I really shouldn't.

File: tests/training/metrics_test.py [hunk 11]
```diff
@@ -211,3 +213,89 @@ def test_f1_measure_works_for_sequences(self):
...
+
+        o_index = self.vocab.get_token_index("O", "tags")
+        tag_vocab = self.vocab.get_index_to_token_vocabulary("tags")
+        metric = ConllSpanBasedF1Measure(tag_vocab, ignore_classes=[o_index])
+
+        metric(prediction_tensor, gold_tensor)
+
+        assert metric._true_positives["ARG1"] == 1
+        assert metric._true_positives["ARG2"] == 0
+        assert metric._true_positives["O"] == 0
```
Comment (matt-gardner) [inline tests/training/metrics_test.py]: I _think_ you should never get an "O" key here at all, right?

Comment (DeNeutoy) [inline tests/training/metrics_test.py]: Yes, that's correct - this was my poor man's way of checking that, just by checking it's the default value of the defaultdict.

File: allennlp/training/metrics.py (hunk 6 above)
Comment (DeNeutoy) [inline allennlp/training/metrics.py]: Great idea, this is much better.

File: allennlp/training/metrics.py (hunk 2 above)
Comment (DeNeutoy) [inline allennlp/training/metrics.py]: I wanted to convert the tensors to lists in the loop, just to make sure nothing weird happened - perhaps this isn't necessary and we could just do this on the tensor directly. 
You do have to do this though, as the `torch.cuda` api is not exactly the same - e.g you cannot call `.tolist()` on a cuda tensor (or possibly `Variable`), I ran into this when computing the losses previously.

File: allennlp/models/semantic_role_labeler.py (hunk 7 above)
Comment (DeNeutoy) [inline allennlp/models/semantic_role_labeler.py]: Done - I also added `Dict[str, float]` to the base class. I think it's impractical to pass this as anything other than a dict, really.

**Comment by DeNeutoy** [comment]
I've made the changes, thanks for the suggestions. The only one I didn't do was building it from params - that seemed a bit ugly, as passing the whole vocab seemed a bit wasteful and I thought you might have a better idea. Maybe that is the best way...

File: tests/training/metrics_test.py (hunk 11 above)
Comment (matt-gardner) [inline tests/training/metrics_test.py]: `assert "O" not in metric._true_positives`

File: tests/training/metrics_test.py [hunk 12]
```diff
@@ -211,3 +213,88 @@ def test_f1_measure_works_for_sequences(self):
...
+        vocab.add_token_to_namespace("O", "tags")
+        vocab.add_token_to_namespace("B-ARG1", "tags")
+        vocab.add_token_to_namespace("I-ARG1", "tags")
+        vocab.add_token_to_namespace("B-ARG2", "tags")
+        vocab.add_token_to_namespace("I-ARG2", "tags")
+        self.vocab = vocab
+
+    def test_conll_span_based_f1_extracts_correct_spans(self):
+        tag_vocab = self.vocab.get_index_to_token_vocabulary("tags")
+        metric = SpanBasedF1Measure(tag_vocab, ignore_classes=["O"])
```
Comment (matt-gardner) [inline tests/training/metrics_test.py]: There should be a case here for when an actual span label is ignored, and again ignoring "O" does nothing now.

**Comment by DeNeutoy** [comment]
@matt-gardner, ready for another look.

File: allennlp/models/semantic_role_labeler.py [hunk 13]
```diff
@@ -175,6 +180,16 @@ def tag(self, text_field: TextField, verb_indicator: IndexField) -> Dict[str, An
 
         return {"tags": tags, "class_probabilities": predictions.numpy()}
 
+    def get_metrics(self, reset: bool = False):
+        metric_dict = self.span_metric.get_metric(reset=reset)
+        if self.training:
+            # This can be a lot of metrics, as there are 3 per class.
+            # During training, we only really care about the overall
+            # metrics, so we filter for them here.
+            return {x: y for x, y in metric_dict if "overall" in x}
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeler.py]: I don't think this is actually going to work - you'll get different metrics for validation and train, and so the training code will crash.

Hmm...  maybe not, because the training code uses the keys in the training metric dict to print out both validation and training metrics...  Ok.  I don't think this is ideal, but the other solutions are ugly too, and this is probably good enough for now.  This could very easily break in the future, though.
